tests/test_references.py

```diff
@@ -186,13 +215,303 @@ async def test_deploy_end_to_end(
     assert [msg for msg in result.result["data"] if "Observed value: {'inner.something': 'testx'}" in msg["msg"]]
 
 
+def test_references_in_plugins(snippetcompiler: "SnippetCompilationTest", modules_v2_dir: str) -> None:
+    """
+    Verify the validation of references in plugins, both on the boundary, and on access through a proxy.
+    """
+    refs_module = os.path.join(modules_v2_dir, "refs")
+
+    # set up project
+    snippetcompiler.setup_for_snippet(
+        "import refs",
+        install_v2_modules=[env.LocalPackagePath(path=refs_module)],
+        autostd=True,
+    )
+
+    def run_snippet(snippet: str) -> None:
+        """
+        Wrapper around snippetcompiler.setup_for_snippet + runs compile and export.
+        Passes appropriate options and prepends snippet with refs import.
+        """
+        snippetcompiler.setup_for_snippet(f"import refs import refs::dc import refs::plugins\n{snippet}", install_project=False, autostd=True)
+        snippetcompiler.do_export()
+
+    # Primitives
+
+    # Scenario: plugin argument annotated as `object`
+    run_snippet(snippet="refs::plugins::takes_obj('hello')")
+    with pytest.raises(PluginTypeException, match="is a reference"):
+        run_snippet(snippet="refs::plugins::takes_obj(refs::create_string_reference('name'))")
+    ## accepts list with a reference in it and even allows access -> infeasible to check inside a black box
+    ## -> this is not a strong requirement. The assertion is here simply to ensure the behavior remains stable
+    run_snippet(snippet="refs::plugins::takes_obj(['hello', refs::create_string_reference('hello')])")
+    run_snippet(snippet="refs::plugins::iterates_obj(['hello', refs::create_string_reference('hello')])")
```

Comment:
Isn't this supposed to fail at the moment where the iterator tries to wrap the reference?

Reply:
Ideally it would be, but this is a scenario that was deemed an acceptable "false negative", where we allow the reference to slip through the mazes. Because it's non-trivial to check, given the current state of the validation and proxy logic, and it's an unlikely scenario, that is still technically correct when it comes to the Python typing.

Reply:
Does this mean that this list is not wrapped in a DynamicProxy? 

i.e. it can leak compiler internals?